etc/ha_healthchecker/ha_healthchecker.py

```diff
@@ -0,0 +1,748 @@
+            # if using status command, 0 means active, non-zero means error
+            # status.
+            subprocess.check_output(cmd.split(), stderr=subprocess.STDOUT)
+            LOG.debug("Run CMD: %(cmd)s", {'cmd': cmd})
+            return 'up'
+        except subprocess.CalledProcessError as e:
+            LOG.error("Failed to check %(srvc)s service: "
+                      "%(err)s %(out)s", {'srvc': service, 'err': e,
+                                          'out': e.output})
+            return 'down'
+
+
+class Refresher(Base):
+    def __init__(self, zk, heartbeat_timeout):
+        super(Refresher, self).__init__(zk, heartbeat_timeout)
+
+    def _local_node_service_process(self, node_obj):
+        service_objs = self.zk.list_services(node_name_filter=node_obj.name)
+        for service_obj in service_objs:
+            self._refresh_service(service_obj, node_obj)
+
+        self._report_heart_beat(node_obj)
+        if node_obj.alarmed and node_obj.role == 'slave':
+            self.zk.update_node(node_obj.name, alarmed=False)
+
+    def _refresh_service(self, service_obj, node_obj):
+        cur_status = self._get_service_status(service_obj.name)
+        LOG.debug("Service %(name)s check result: %(status)s",
+                  {'name': service_obj.name, 'status': cur_status})
+        update_dict = {}
+        if cur_status == 'up':
+            if service_obj.status != 'up':
+                update_dict['status'] = 'up'
+                update_dict['restarted'] = False
+                update_dict['alarmed'] = False
+        else:
+            if not service_obj.restarted:
+                update_dict['status'] = 'restarting'
+                update_dict['restarted'] = True
+            else:
+                update_dict['status'] = 'down'
+        if update_dict:
+            self.zk.update_service(service_obj.name, node_obj.name,
+                                   **update_dict)
+
+        if cur_status != 'up':
+            LOG.debug(
+                "Service %(name)s is abnormal. It's status is %(status)s.",
+                {'name': service_obj.name,
+                 'status': cur_status.upper()})
+
+    def _report_heart_beat(self, node_obj):
+        hb = datetime.datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')
+        update_dict = {'heartbeat': hb}
+        if node_obj.status == 'initializing':
+            update_dict['status'] = 'up'
+        self.zk.update_node(node_obj.name, **update_dict)
+        LOG.debug("Report node %(name)s heartbeat %(hb)s",
+                  {'name': node_obj.name, 'hb': hb})
+
+    def _oppo_node_check(self, oppo_node_obj):
+        if self.node.type == 'zookeeper':
+            return
+        if oppo_node_obj.status == 'maintaining':
+            return
+        if (not self._ping(oppo_node_obj.ip) and
+                self._is_check_heart_beat_overtime(oppo_node_obj)):
+            if oppo_node_obj.role == 'master' and oppo_node_obj.status == 'up':
```

Comment:
no matter what role the oppo_node is, we should mark it to `down`,  right?

Reply:
As we treat 'down' 'maintainning' for skip process. It's reasonable to set master node to down, as the host is done. We can use the down status to check whether we need to do the switch. For slave node, what we want to do is just report a issue, as we only care about master to slave switch. If before the switch we set the oppo slave node to down, we may skip so many process. Also we need a way to set it back, that why the L151 to L152, I only specify the alarmed slave node.